File: BurrowsWheelerTransform/cpp/bwt.cpp
// Burrows-Wheeler Transform cpp soln by Steven A Dunn for Code Eval

#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::sort;
using std::string;
using std::vector;

vector<string> split(const string&);

int main (int argc, char* argv[])
{
    ifstream input_file(argv[1]);
    string line;

    if (input_file)
    {
        while (getline(input_file, line))
        {
            line = line.substr(0, line.find("|"));
            int N = line.size();
            vector<string> result(N, "");
            vector<string> sorted_result(N, "");
            vector<string> entry = split(line);

            for (int i = 0; i < N; ++i)
            {
                for (int i = 0; i < result.size(); ++i)
                {
                    int len = entry[i].size();
                    result[i] += entry[i][len-1];
                }
                sorted_result = result;
                sort(sorted_result.begin(), sorted_result.end());
                entry = sorted_result;
            }
            for (int i = 0; i < result.size(); ++i)
            {
                if (result[i].find("$") == result[i].size() - 1)
                {
                    cout << result[i] << endl;
                    break;
                }
            }
        }
        input_file.close();
    }
    return 0;
}

vector<string> split(const string& line)
{
    vector<string> split_line;
    for (int i = 0; i < line.size(); ++i)
        split_line.push_back(string(1, line[i]));
    return split_line;
}

File: BurrowsWheelerTransform/py2/bwt.py
# Burrows-Wheeler Transform py2 soln by Steven A Dunn for Code Eval

import sys

def add(result, entry):
    for i in range(0, len(result)):
        result[i] += entry[i][-1]
    return result

f = open(sys.argv[1], 'r')
for line in f:
    line = line.strip()[0:-1]
    entry = list(line)
    N = len(entry)
    result = [""] * N
    sorted_result = [""] * N
    for i in range(0, N):
        result = add(result, entry)
        sorted_result = result[:]
        sorted_result.sort()
        entry = sorted_result
    for value in result:
        if value[-1] == "$":
            print value
            break
f.close()